src/type/result.ts

```diff
@@ -0,0 +1,74 @@
+export type Result<T, E extends string = string> =
+  | SuccessResult<T>
+  | FailureResult<E>;
+export const success = <T>(value: T): SuccessResult<T> => ({ ok: true, value });
+export type SuccessResult<T> = { ok: true; value: T };
+type Failure = {
+  (error: Error): FailureResult<typeof unknownCode>;
+  <C extends string | undefined>(code: C, error: Error): FailureResult<
+    C extends string ? C : typeof unknownCode
+  >;
+};
+const unknownCode = "UNKNOWN" as const;
+export const failure: Failure = <C extends string>(
+  codeOrError: C | undefined | Error,
+  error?: Error
+) => {
+  switch (typeof codeOrError) {
+    case "undefined": {
+      if (typeof error === "undefined") {
+        throw new Error("Error must be specified");
+      }
+
+      return {
+        ok: false as const,
+        code: unknownCode,
+        error: error,
+      };
+    }
+    case "string": {
+      if (error === undefined) {
+        throw new Error("Error must be specified");
+      }
+
+      return { ok: false as const, code: codeOrError, error };
+    }
+    case "object": {
+      if (codeOrError === null) {
```

Comment:
バグってますね！　やっぱnullとundefinedは片方使用不可にしたほうがいい気がしますね！！
```suggestion
      if (error === undefined) {
```

Reply:
記憶が正しければ：
nullはobjectなのでここで弾いてた記憶があります…が、Copilotが余計なお世話してる説もあるので消して試してみます

Reply:
あーーー　なるほどです。
switch使うとErrorを`object`として扱わないといけなくなるので、普通にif elseにするのどうでしょう。
undefinedだったときとstringだったときは同じ処理になるので、なんかこんな感じで書けるかなとか。

```ts
  if (codeOrError instanceof Error) {
    return {
      ok: false as const,
      code: undefined,
      error: codeOrError,
    };
  } else if (codeOrError == undefined || typeof codeOrError === "string") {
    if (error === undefined) {
      throw new Error("Error must be specified");
    }
    return { ok: false as const, code: codeOrError, error };
  } else {
    throw new Error("Error must be specified");
  }
```

Reply:
消しても動きました。